firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java

```diff
@@ -655,15 +663,26 @@ private Bound boundFromFields(String methodName, Object[] values, boolean before
                   + ".");
         }
         String documentId = (String) rawValue;
-        if (documentId.contains("/")) {
+        if (!this.query.isCollectionGroupQuery() && documentId.contains("/")) {
           throw new IllegalArgumentException(
-              "Invalid query. Document ID '"
+              "Invalid query. When querying a collection and ordering by FieldPath.documentId(), "
+                  + "the value passed to "
+                  + methodName
+                  + "() must be a plain document ID, but '"
                   + documentId
-                  + "' contains a slash in "
+                  + "' contains a slash.");
+        }
+        ResourcePath path = query.getPath().append(ResourcePath.fromString(documentId));
+        if (!DocumentKey.isDocumentKey(path)) {
+          throw new IllegalArgumentException(
+              "Invalid query. When querying a collection group and ordering by "
+                  + "FieldPath.documentId(), the value passed to "
                   + methodName
-                  + "().");
+                  + "() must result in a valid document path, but '"
+                  + path
```

Comment:
So these errors are actually redundant.  I could get rid of the first one, but I think it will be hit more often, and by handling it separately we can make the error message a little clearer.  That said, I did go back and forth on this. I'll think about it a bit more as I do the iOS port.